File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        // Start with middle value

        // If middle value is lower, exclude it and take half again and go to next value to check. 

        // If middle value is higher, exclude and take remaining lower half and divide by 2 to check next value. 

        // Keep doing this process until value is found or cannot divide by 2, meaning not foundabstract

        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            
            int middle = left + (right - left) / 2;

            if (nums[middle] == target) {
                return middle;
            }

            if (nums[middle] > target) {
                right = middle - 1;

            } else {
                left = middle + 1;
            }
        }

        return -1;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-0.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        columns = len(matrix[0])

        top = 0
        bottom = rows - 1
        while top <= bottom:
            row = (top + bottom) // 2
            if target > matrix[row][-1]:
                top = row + 1
            elif target < matrix[row][0]:
                bottom = row - 1
            else: 
                break

        if not (top <= bottom):
            return False

        row = (top + bottom) //2
        left = 0
        right = len(matrix[0]) - 1    

        while left <= right:
            middle = (left + right) // 2

            if target > matrix[row][middle]:
                left = middle + 1

            elif target < matrix[row][middle]:
                right = middle - 1

            else:
                return True

        return False

            
            